### PR TITLE
[WIP] Time-major layout for RNN

### PR DESCRIPTION
**Pull request by pluskid**
@piiswrong It seems that you modified some shape related code of `SoftmaxOutput` in 2a9ca98e3ac12fb48848a4f88c29260f66f250b4 and a717570e3ea49f0761fef3779d88e0186f8dd6d5. It somehow broke our current speech demos (cc @antinucleon @yzhang87), and also I'm having a hard time to get the time-major layout RNN to work here. Could you give us a quick summary of what those modifications are suppose to do?

Specifically, it seems [this block](https://github.com/dmlc/mxnet/commit/2a9ca98e3ac12fb48848a4f88c29260f66f250b4#diff-772857f6d2e364650368f86ab6cdf11aL85) of modification is causing the issue. The original behavior of `SoftmaxOutput` is treating only the last dimension as the *class* dimension. For example, if the input is of shape `(2, 3, 4)`, then the output is also `(2, 3, 4)`, while the label is of shape `(2, 3)`, as `4` here means 4 categories. This is also what the current `infer_shape` says.

However, if I understand correctly, it seems that modification is forcing the first dimension to be *number of samples* and everything else is collapsed as categories. Is this intended behavior that is needed in some specific application? If it is, I guess we need to use a switch as we need the original behavior to properly support time-major RNNs.

### CODE REVIEW

**Comment by piiswrong** [comment]
I changed it because the previous behavior didn't make sense for image data. You can add a new mode for RNN

**Comment by pluskid** [comment]
@piiswrong Can you add some doc about what is the current expected behavior? `(a, b, c, d)` tensor is treated as `(a, b*c*d)` tensor? 

My understanding is that for image classifications, the last layer is usually already flatten in 2D, or for segmentation is in multi-output mode. It would be better if we could find a way to uniformly handle image and RNN, instead of adding another mode.

**Comment by piiswrong** [comment]
(n,c*h*w) can be useful for attention.
Although (n, c, h*w) also works. So i guess we can use (n, c, h*w) if you don't want one more mode

File: example/rnn-time-major/rnn_cell_demo.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                                  time_major=True)
+
+    def sym_gen(seq_len):
+        data = mx.sym.Variable('data')
+        label = mx.sym.Variable('softmax_label')
+        embed = mx.sym.Embedding(data=data, input_dim=len(vocab),
+                                 output_dim=num_embed, name='embed')
+
+        # TODO(tofix)
+        # currently all the LSTM parameters are concatenated as
```
Comment (piiswrong) [inline example/rnn-time-major/rnn_cell_demo.py]: add a new initialization method

File: include/mxnet/ndarray.h [hunk 2]
```diff
@@ -394,6 +394,18 @@ void CopySliceTo(const NDArray &from, int slice_dim, index_t start, index_t end,
...
 /*!
+ * \brief assign a slice along any axis.
+ * \param from the NDArray whose value is used for assigning
+ * \param slice_dim the axis we want to perform slice in
+ * \param start the beginning of the slice
+ * \param end the ending of the slice
+ * \param to the bigger NDArray whose slice we want to assign to
+ * \param priority the priority of the task
+ */
+void AssignSliceFrom(const NDArray &from, int slice_dim, index_t start, index_t end,
```
Comment (piiswrong) [inline include/mxnet/ndarray.h]: write an ndarray op instead of a c api

File: python/mxnet/ndarray.py [hunk 3]
```diff
@@ -303,6 +303,23 @@ def _copy_slice_to(self, axis, start, stop, target):
         assert isinstance(target, NDArray)
         return _internal._copy_slice_to(self, axis, start, stop, out=target)
 
+    def assign_slice_from(self, axis, start, stop, source):
```
Comment (piiswrong) [inline python/mxnet/ndarray.py]: remove this. use crop & a new partial set op

Comment (pluskid) [inline python/mxnet/ndarray.py]: Do you mean creating a new op or are you referring to some existing op that I missed?

File: include/mxnet/ndarray.h (hunk 2 above)
Comment (pluskid) [inline include/mxnet/ndarray.h]: Yes, this is actually an ndarray op.

**Comment by pluskid** [comment]
Closed for #3497

**Pull request closed without merging**